File: anny_dang/homework4/boggle.py
def boggle(dictionary, board):
    """
    idea:
    - create dictionary trie
        - walk through each word in dictionary
        - if len(word) < 3:
            continue
        - convert word to lowercase
        - have cur = trie
        - for each character in word:
            - if character not in cur:
                - add it
            - move cur to that character node
        - at the end of word:
            - store full word at key "KEY"

    - backtracking function
        - base cases:
            - if cell out of board -> return
            - if board cell is "#" -> return
            - convert board character to lowercase
            - if character not in current trie node -> return
        - move to next trie node
        - if "KEY" in that node:
            - add that word to ans
            - delete "KEY" so we do not add duplicate again
        - mark current board cell as visited with "#"
        - walk through all 8 directions:
            - call backtrack on neighbor cell with next trie node
        - change board cell back to original character

    - main loop:
        - walk through each cell in board
        - if lowercase character at board cell is in trie:
            - call backtrack(cell, trie)
    
    n: number of all characters, L: max length of a word
    time: O(n)
    space: O(n + L)
    """
    trie = {}
    ans = []
    directions = [(0, 1), (1, 0), (-1, 0), (0, -1), (1, -1), (-1, 1), (1, 1), (-1, -1)]
    for word in dictionary:
        if len(word) < 3:
            continue

        lower_word = word.lower()
        cur = trie
        for ch in lower_word:
            if ch not in cur:
                cur[ch] = {}
            cur = cur[ch]
        
        cur["KEY"] = word

    def backtrack(r, c, node):
        if r < 0 or r >= len(board) or c < 0 or c >= len(board[0]):
            return 

        if board[r][c] == "#":
            return 
        
        ch = board[r][c].lower()
        if ch not in node:
            return 
        
        next_node = node[ch]

        if "KEY" in next_node:
            ans.append(next_node["KEY"])
            del next_node["KEY"]
        
        temp = board[r][c]
        board[r][c] = "#"
        for x, y in directions:
            backtrack(r + x, c + y, next_node)
        
        board[r][c] = temp
    
    for r in range(len(board)):
        for c in range(len(board[0])):
            if board[r][c].lower() in trie:
                backtrack(r, c, trie)
    
    return ans 


if __name__ == "__main__":
    dictionary = [
        "Ace", "Ape", "Cape", "Clap", "Clay",
        "Gape", "Grape", "Lace", "Lap", "Lay",
        "Mace", "Map", "May", "Pace", "Pay",
        "Race", "Rap", "Ray", "Tap", "Tape",
        "Trace", "Trap", "Tray", "Yap",
    ]

    board = [
        ["A", "D", "E"],
        ["R", "C", "P"],
        ["L", "A", "Y"],
    ]

    expected = [
        "Ace", "Race", "Pace", "Lace", "Pay",
        "Lay", "Clay", "Ray", "Lap", "Rap",
        "Clap", "Ape", "Cape", "Yap",
    ]

    print("Dictionary:")
    print(dictionary)
    print()

    print("Board:")
    for row in board:
        print(row)
    print()

    print("Expected Output:")
    print(expected)
    print()

    print("Your Output:")
    print(boggle(dictionary, board))

File: Part2Problems/Boggle.java
package Part2Problems;
import Part1DataStructureImplementation.Trie;
import java.util.*;

//time spend 35min 
//time and space i am not sure im confused, i think o(n) for both ik recursions tack is o(n)
public class Boggle {
    Trie trie = new Trie();
    public List<String> boggle(Set<String> dictionary, char[][] board){
       
        for(String word: dictionary){
            trie.insert(word.toLowerCase());
        }
        boolean[][] visited = new boolean[board.length][board[0].length];
        Set<String> currSet = new HashSet<>();
        for(int i=0;i<board.length;i++){
            for(int j=0;j<board[0].length;j++){
                
                dfs(board,visited,i,j,"" ,currSet);
                
            }
        }
        return new ArrayList<>(currSet);
    }
    private void dfs(char[][] board, boolean[][] visited,int row,int col,String currWord,Set<String> seenSet){
        if(row<0 || col <0 || row>= board.length || col >=board[0].length|| visited[row][col]){
            return;
        }
        currWord += Character.toLowerCase(board[row][col]);
        if(!trie.startsWith(currWord)){
            return;
        }
        if (trie.isValidWord(currWord)) {
    seenSet.add(currWord);
}

        visited[row][col] = true;
        dfs( board,  visited, row+1, col,currWord, seenSet);
        dfs( board,  visited, row-1, col,currWord, seenSet);
        dfs( board,  visited, row, col+1,currWord, seenSet);
        dfs( board,  visited, row, col-1,currWord, seenSet);
        dfs( board,  visited, row+1, col+1,currWord, seenSet);
        dfs( board,  visited, row-1, col-1,currWord, seenSet);
         dfs( board,  visited, row+1, col-1,currWord, seenSet);
          dfs( board,  visited, row-1, col+1,currWord, seenSet);
        visited[row][col] = false;
      
    }
    public static void main(String[] args) {
    Boggle boggle = new Boggle();

    Set<String> dictionary = new HashSet<>();

    dictionary.add("Ace");
    dictionary.add("Ape");
    dictionary.add("Cape");
    dictionary.add("Clap");
    dictionary.add("Clay");
    dictionary.add("Race");
    dictionary.add("Ray");
    dictionary.add("Tap");
    dictionary.add("Tape");
    dictionary.add("Trace");
    dictionary.add("Trap");
    dictionary.add("Tray");
    dictionary.add("Yap");

    char[][] board = {
        {'A', 'D', 'E'},
        {'R', 'C', 'P'},
        {'L', 'A', 'Y'}
    };

    List<String> result = boggle.boggle(dictionary, board);

    System.out.println("Words found:");
    for (String word : result) {
        System.out.println(word);
    }
}
}
        
